Provide for Yarn/MapReduce memory allocation (Hadoop)

renevatium opened the pull request:
After successfully installing the Analytics Stack to a single server (Ubuntu 12.04 on t1.medium) my remote-tasks were constantly running out of memory (t1.medium has less the 2GB). Neither increasing swapfile size nor setting lower/higher swappiness made any difference. Found yarn/mapred settings and applied manually to mapred-site.xml and yarn-site.xml respectively. Was then able to run remote tasks. 

This update provides for configuration of memory allocation. Values have been set to the standard defaults. I have purposefully limited the config to the necessary vars as there are just too many options to allow for unrestricted configuration. Just trying to increase automation.

edx-webhook commented:
Thanks for the pull request, @renevatium! I've created [OSPR-1151](https://openedx.atlassian.net/browse/OSPR-1151) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/configuration/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

nedbat commented:
@renevatium Thanks, I've assigned this to the devops team.

mulby commented on playbooks/roles/hadoop_common/templates/mapred-site.xml.j2:
```diff
@@ -6,4 +6,36 @@
+  </property>
+  <property>
+    <name>mapreduce.map.java.opts</name>
+    <value>-Xmx{{ mapreduce_map_memory_mb }}M</value>
```
This thread suggests that this value should actually be smaller than the one above:

http://stackoverflow.com/questions/24070557/what-is-the-relation-between-mapreduce-map-memory-mb-and-mapred-map-child-jav

renevatium replied:
Agreed. Was working off an incorrect assumption. Looks like the default java.opts value is closer to a fifth of memory allocation. Will update the branch to define opts independently and set to Hadoop defaults.

mulby replied:
Sounds good. Ping me directly on this PR again when this is ready for review again.

renevatium commented:
@mulby Sorry for delay, in the middle of launch. Have updated the defaults to AWS [recommended](http://docs.aws.amazon.com/ElasticMapReduce/latest/DeveloperGuide/TaskConfiguration_H2.html) for task config (m1.medium). Added Yarn scheduler vars. Removed task.io settings as not necessary and low values there were causing build issues. R

mulby commented on playbooks/roles/hadoop_common/defaults/main.yml:
```diff
@@ -60,3 +60,19 @@ hadoop_common_debian_pkgs:
 hadoop_common_redhat_pkgs: []
+
+#
+# MapReduce/Yarn memory config (defaults for m1.medium)
```
Many of our users of these playbooks are running them on Vagrant boxes for development purposes as well.

Could we make these settings opt-in? That is: use them if they are defined, but omit them from the config file if they aren't? And leave them undefined by default?

renevatium commented:
@mulby Have reconfigured as optional, conditional vars. Have tested in a release environment with the defaults as per comment in [main.yml](https://github.com/proversity-org/configuration/blob/config/mapred/playbooks/roles/hadoop_common/defaults/main.yml#L65) file. Passed values as $EXTRA_VARS to analytics_single playbook. Will put together a doc this weekend. Another benefit of this method is that vars are not limited, any yarn-site or mapred-site settings can be applied. R

mulby commented:
I love this, big improvement.

@edx/devops - can one of you review this PR from an ansible patterns standpoint?

:+1:

mulby commented:
also - @renevatium can you squash and rebase this change in preparation for merge?

renevatium commented:
@mulby never actually had to do a squash before, might have gotten it confused so let me know if I need to push again. Squashed and Rebased to /edx/configuration/master.

mulby commented:
heh, yeah I think you did the squash incorrectly

This PR should just have one commit on it.

renevatium commented:
@mulby Still not 100% sure but seems right. Let me know if you need anything else.

mulby commented:
@renevatium looks good now. Thanks, :+1:

I'm just waiting on a quick review from our ops team before merging this.

@jibsheet - can you take a quick look at this before I merge it?

feanil commented:
:+1:
